File: frontend/lib/surface.ts
/**
 * Tilo Surface Protocol v1 — TypeScript types.
 *
 * Hand-aligned with `backend/app/schemas/surface.py` and the JSON Schema
 * exported at `frontend/lib/surface.schema.json`. The latter is the
 * machine-readable source of truth; this file mirrors it for ergonomic
 * use in TypeScript renderers. Drift is caught by the protocol round-trip
 * tests and the schema-export `--check` mode (see `scripts/export_surface_schema.py`).
 */

export const SURFACE_SCHEMA_VERSION = "tilo.surface.v1" as const;

export const NOOP_ACTION_ID = "__noop__" as const;

export type SurfaceIntent =
  | "request_approval"
  | "collect_input"
  | "present_result"
  | "offer_choices"
  | "confirm_memory"
  | "show_progress"
  | "escalate_to_rich"
  | "ask_clarification";

export type BudgetHint = "mini" | "rich";
export type BlockCompat = "graceful" | "strict";

export type Severity = "info" | "low" | "medium" | "high";

export type SurfaceBlockType =
  | "heading"
  | "text"
  | "evidence"
  | "comparison"
  | "decision"
  | "form"
  | "progress"
  | "list"
  | "link"
  | "editable"
  | "artifact_link"
  | "fallback";

// ---------- Block data shapes (mirror Pydantic models) -------------------- //

export interface HeadingData { text: string; severity?: Severity | null }
export interface TextData { content: string }
export interface EvidenceData { excerpt: string; source_ref: string; source_label?: string | null }

export interface ComparisonSide { label: string; value: string; severity?: Severity | null }
export interface ComparisonRow { label: string; left: string; right: string; severity?: Severity | null }
export interface ComparisonData {
  shape: "side_by_side" | "table";
  left?: ComparisonSide | null;
  right?: ComparisonSide | null;
  rows: ComparisonRow[];
}

export interface DecisionOption {
  id: string;
  label: string;
  value: string;
  action_id: string;
  severity?: Severity | null;
}
export interface DecisionData {
  prompt?: string | null;
  mode: "single" | "multi";
  options: DecisionOption[];
}

export type FormFieldKind = "text" | "number" | "textarea" | "select" | "toggle" | "date";
export interface FormFieldOption { label: string; value: string }
export interface FormField {
  name: string;
  label: string;
  kind: FormFieldKind;
  required: boolean;
  placeholder?: string | null;
  min?: number | null;
  max?: number | null;
  step?: number | null;
  options: FormFieldOption[];
  default?: unknown;
}
export interface FormData { fields: FormField[]; submit_action_id: string }

export type ProgressShape = "steps" | "percent" | "status";
export type ProgressState = "pending" | "running" | "done" | "failed" | "skipped";
export interface ProgressStep { id: string; label: string; state: ProgressState }
export interface ProgressData {
  shape: ProgressShape;
  percent?: number | null;
  status?: string | null;
  steps: ProgressStep[];
}

export interface ListItemData { text: string; severity?: Severity | null }
export interface ListData { ordered: boolean; items: ListItemData[] }

export type LinkTarget = "drawer" | "page" | "webview" | "external";
export interface LinkData { label: string; url: string; target: LinkTarget }

export type EditableKind = "rich_text" | "structured";
export interface EditableData {
  kind: EditableKind;
  value: string;
  schema?: Record<string, unknown> | null;
  submit_action_id: string;
  highlights: string[];
}

export interface ArtifactLinkData {
  artifact_id: string;
  title: string;
  summary?: string | null;
  open_action_id: string;
}

export interface FallbackData { content: string }

// ---------- Action contract (unchanged from artifact_spec.v1) ------------- //

export type SurfaceActionType =
  | "approve" | "reject" | "edit" | "select"
  | "continue_task" | "regenerate" | "invoke_tool"
  | "create_memory" | "promote_skill" | "export" | "confirm";

export interface SurfaceStateBinding {
  entity_type:
    | "artifact" | "confirmation" | "memory"
    | "skill_candidate" | "tool_invocation" | "task" | "run";
  entity_id: string;
  field?: string | null;
}

export interface SurfaceAction {
  id: string;
  label: string;
  action_type: SurfaceActionType;
  confirmation_required: boolean;
  confirmation_id?: string | null;
  payload?: Record<string, unknown>;
  state_binding?: SurfaceStateBinding | null;
}

// ---------- Block envelope ------------------------------------------------ //

interface BlockEnvelope<T extends SurfaceBlockType, D> {
  id: string;
  type: T;
  data: D;
  fallback_text: string;
  actions?: SurfaceAction[];
  state_binding?: SurfaceStateBinding | null;
}

export type SurfaceBlock =
  | BlockEnvelope<"heading", HeadingData>
  | BlockEnvelope<"text", TextData>
  | BlockEnvelope<"evidence", EvidenceData>
  | BlockEnvelope<"comparison", ComparisonData>
  | BlockEnvelope<"decision", DecisionData>
  | BlockEnvelope<"form", FormData>
  | BlockEnvelope<"progress", ProgressData>
  | BlockEnvelope<"list", ListData>
  | BlockEnvelope<"link", LinkData>
  | BlockEnvelope<"editable", EditableData>
  | BlockEnvelope<"artifact_link", ArtifactLinkData>
  | BlockEnvelope<"fallback", FallbackData>;

// ---------- Top-level SurfaceSpec ----------------------------------------- //

export interface ProvenanceRef { type: string; id: string; label?: string | null }

export interface SurfaceFallbacks {
  telegram?: { inline_keyboard: Array<Array<Record<string, unknown>>> };
  slack?: { blocks: Array<Record<string, unknown>> };
  email?: { html: string };
  [channel: string]: unknown;
}

export interface SurfaceSpec {
  schema_version: typeof SURFACE_SCHEMA_VERSION;
  surface_id: string;
  turn_id: string;
  run_id: string;

  intent: SurfaceIntent;
  budget_hint: BudgetHint;
  block_compat: BlockCompat;

  blocks: SurfaceBlock[];
  fallback_text: string;
  fallbacks?: SurfaceFallbacks | null;

  provenance: ProvenanceRef[];
  memory_refs: string[];
  metadata: Record<string, unknown>;
}

// ---------- Server-side row wrapping a SurfaceSpec ------------------------ //

export interface SurfaceTurn {
  id: string;
  run_id: string;
  session_id: string | null;
  workspace_id: string;
  project_id: string | null;
  ordinal: number;
  intent: SurfaceIntent;
  budget_hint: BudgetHint;
  status: string;
  surface_spec_json: SurfaceSpec;
  policy_decision_json: {
    step_index: number;
    decision: string;
    intent: string | null;
    surface: string | null;
    reason: string;
    rule_id: string | null;
  } | null;
  plan_step_index: number | null;
  plan_step_type: string | null;
  artifact_id: string | null;
  composer_mode: "deterministic" | "llm" | "deterministic_fallback" | string;
  created_at: string;
}

File: backend/tilo/schemas/surface.py
"""Tilo Surface Protocol v1 — Pydantic models.

This module is the single source of truth for the Surface Protocol.
The companion JSON Schema (frontend/lib/surface.schema.json) is generated
from these models via tools/export_surface_schema.py.

See docs/SURFACE_PROTOCOL.md for the normative human-readable spec.
"""

from __future__ import annotations

from enum import StrEnum
from typing import Annotated, Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from tilo.schemas.artifact import (
    ArtifactAction,
    ProvenanceRef,
    StateBinding,
    SUPPORTED_ACTION_TYPES,  # re-exported for downstream code
)

SURFACE_SCHEMA_VERSION: Literal["tilo.surface.v1"] = "tilo.surface.v1"


# --------------------------------------------------------------------------- #
# Intent vocabulary (closed)                                                  #
# --------------------------------------------------------------------------- #


class SurfaceIntent(StrEnum):
    request_approval = "request_approval"
    collect_input = "collect_input"
    present_result = "present_result"
    offer_choices = "offer_choices"
    confirm_memory = "confirm_memory"
    show_progress = "show_progress"
    escalate_to_rich = "escalate_to_rich"
    ask_clarification = "ask_clarification"


class BudgetHint(StrEnum):
    mini = "mini"
    rich = "rich"


class BlockCompat(StrEnum):
    graceful = "graceful"
    strict = "strict"


# Reserved option action_id meaning "no backend action; client-side only".
NOOP_ACTION_ID = "__noop__"


# --------------------------------------------------------------------------- #
# Block-type vocabulary (closed)                                              #
# --------------------------------------------------------------------------- #


class SurfaceBlockType(StrEnum):
    heading = "heading"
    text = "text"
    evidence = "evidence"
    comparison = "comparison"
    decision = "decision"
    form = "form"
    progress = "progress"
    list = "list"  # noqa: A003 — closed enum value, intentional
    link = "link"
    editable = "editable"
    artifact_link = "artifact_link"
    fallback = "fallback"


class Severity(StrEnum):
    info = "info"
    low = "low"
    medium = "medium"
    high = "high"


class ProgressState(StrEnum):
    pending = "pending"
    running = "running"
    done = "done"
    failed = "failed"
    skipped = "skipped"


class FormFieldKind(StrEnum):
    text = "text"
    number = "number"
    textarea = "textarea"
    select = "select"
    toggle = "toggle"
    date = "date"


class LinkTarget(StrEnum):
    drawer = "drawer"
    page = "page"
    webview = "webview"
    external = "external"


class EditableKind(StrEnum):
    rich_text = "rich_text"
    structured = "structured"


# --------------------------------------------------------------------------- #
# Per-block data shapes                                                       #
# --------------------------------------------------------------------------- #


class _StrictModel(BaseModel):
    """Disallow unknown fields inside block.data shapes — closes the schema."""

    model_config = ConfigDict(extra="forbid")


class HeadingData(_StrictModel):
    text: str = Field(min_length=1)
    severity: Severity | None = None


class TextData(_StrictModel):
    content: str = Field(min_length=1)


class EvidenceData(_StrictModel):
    excerpt: str = Field(min_length=1)
    source_ref: str = Field(min_length=1)
    source_label: str | None = None


class ComparisonSide(_StrictModel):
    label: str = Field(min_length=1)
    value: str = Field(min_length=1)
    severity: Severity | None = None


class ComparisonRow(_StrictModel):
    label: str = Field(min_length=1)
    left: str
    right: str
    severity: Severity | None = None


class ComparisonData(_StrictModel):
    shape: Literal["side_by_side", "table"]
    left: ComparisonSide | None = None
    right: ComparisonSide | None = None
    rows: list[ComparisonRow] = Field(default_factory=list)

    @model_validator(mode="after")
    def _validate_shape(self) -> "ComparisonData":
        if self.shape == "side_by_side":
            if self.left is None or self.right is None:
                raise ValueError("comparison.side_by_side requires both 'left' and 'right'")
            if self.rows:
                raise ValueError("comparison.side_by_side must not include 'rows'")
        else:  # table
            if self.left is not None or self.right is not None:
                raise ValueError("comparison.table must omit 'left'/'right'")
            if not self.rows:
                raise ValueError("comparison.table requires non-empty 'rows'")
        return self


class DecisionMode(StrEnum):
    single = "single"
    multi = "multi"


class DecisionOption(_StrictModel):
    id: str = Field(min_length=1)
    label: str = Field(min_length=1)
    value: str = Field(min_length=1)
    action_id: str = Field(min_length=1)
    severity: Severity | None = None


class DecisionData(_StrictModel):
    prompt: str | None = None
    mode: DecisionMode = DecisionMode.single
    options: list[DecisionOption] = Field(min_length=1)


class FormFieldOption(_StrictModel):
    label: str = Field(min_length=1)
    value: str = Field(min_length=1)


class FormField(_StrictModel):
    name: str = Field(min_length=1)
    label: str = Field(min_length=1)
    kind: FormFieldKind
    required: bool = False
    placeholder: str | None = None
    min: float | None = None
    max: float | None = None
    step: float | None = None
    options: list[FormFieldOption] = Field(default_factory=list)
    default: Any | None = None

    @model_validator(mode="after")
    def _validate_kind(self) -> "FormField":
        if self.kind == FormFieldKind.select and not self.options:
            raise ValueError("form.select fields require non-empty 'options'")
        if self.kind != FormFieldKind.number and (self.min is not None or self.max is not None or self.step is not None):
            raise ValueError("min/max/step are only valid on form.number fields")
        return self


class FormData(_StrictModel):
    fields: list[FormField] = Field(min_length=1)
    submit_action_id: str = Field(min_length=1)


class ProgressShape(StrEnum):
    steps = "steps"
    percent = "percent"
    status = "status"


class ProgressStep(_StrictModel):
    id: str = Field(min_length=1)
    label: str = Field(min_length=1)
    state: ProgressState


class ProgressData(_StrictModel):
    shape: ProgressShape
    percent: int | None = Field(default=None, ge=0, le=100)
    status: str | None = None
    steps: list[ProgressStep] = Field(default_factory=list)

    @model_validator(mode="after")
    def _validate_shape(self) -> "ProgressData":
        if self.shape == ProgressShape.steps and not self.steps:
            raise ValueError("progress.steps requires non-empty 'steps'")
        if self.shape == ProgressShape.percent and self.percent is None:
            raise ValueError("progress.percent requires 'percent'")
        if self.shape == ProgressShape.status and not self.status:
            raise ValueError("progress.status requires 'status'")
        return self


class ListItem(_StrictModel):
    text: str = Field(min_length=1)
    severity: Severity | None = None


class ListData(_StrictModel):
    ordered: bool = False
    items: list[ListItem] = Field(min_length=1)


class LinkData(_StrictModel):
    label: str = Field(min_length=1)
    url: str = Field(min_length=1)
    target: LinkTarget = LinkTarget.external


class EditableData(_StrictModel):
    kind: EditableKind
    value: str
    schema_: dict[str, Any] | None = Field(default=None, alias="schema")
    submit_action_id: str = Field(min_length=1)
    highlights: list[str] = Field(default_factory=list)

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    @model_validator(mode="after")
    def _validate_kind(self) -> "EditableData":
        if self.kind == EditableKind.structured and self.schema_ is None:
            raise ValueError("editable.structured requires 'schema'")
        return self


class ArtifactLinkData(_StrictModel):
    artifact_id: str = Field(min_length=1)
    title: str = Field(min_length=1)
    summary: str | None = None
    open_action_id: str = Field(min_length=1)


class FallbackData(_StrictModel):
    content: str = Field(min_length=1)


# --------------------------------------------------------------------------- #
# Block envelope                                                              #
# --------------------------------------------------------------------------- #


_BLOCK_TYPE_TO_DATA: dict[SurfaceBlockType, type[_StrictModel]] = {
    SurfaceBlockType.heading: HeadingData,
    SurfaceBlockType.text: TextData,
    SurfaceBlockType.evidence: EvidenceData,
    SurfaceBlockType.comparison: ComparisonData,
    SurfaceBlockType.decision: DecisionData,
    SurfaceBlockType.form: FormData,
    SurfaceBlockType.progress: ProgressData,
    SurfaceBlockType.list: ListData,
    SurfaceBlockType.link: LinkData,
    SurfaceBlockType.editable: EditableData,
    SurfaceBlockType.artifact_link: ArtifactLinkData,
    SurfaceBlockType.fallback: FallbackData,
}


class SurfaceBlock(BaseModel):
    """One block in a SurfaceSpec.

    `data` is validated against the type-specific shape in `_BLOCK_TYPE_TO_DATA`.
    On validation, `data` is normalized to the *parsed* model dump so downstream
    code can rely on shape stability.
    """

    id: str = Field(min_length=1)
    type: SurfaceBlockType
    data: dict[str, Any] = Field(default_factory=dict)
    fallback_text: str = Field(min_length=1)
    actions: list[ArtifactAction] = Field(default_factory=list)
    state_binding: StateBinding | None = None

    model_config = ConfigDict(use_enum_values=False)

    @model_validator(mode="after")
    def _validate_data_shape(self) -> "SurfaceBlock":
        shape_cls = _BLOCK_TYPE_TO_DATA[self.type]
        parsed = shape_cls.model_validate(self.data)
        self.data = parsed.model_dump(by_alias=True, exclude_none=False)
        return self

    # Convenient parsed-data accessor
    def parsed_data(self) -> _StrictModel:
        return _BLOCK_TYPE_TO_DATA[self.type].model_validate(self.data)


# --------------------------------------------------------------------------- #
# Channel-fallback hints                                                      #
# --------------------------------------------------------------------------- #


class TelegramFallback(BaseModel):
    inline_keyboard: list[list[dict[str, Any]]] = Field(default_factory=list)


class SlackFallback(BaseModel):
    blocks: list[dict[str, Any]] = Field(default_factory=list)


class EmailFallback(BaseModel):
    html: str = Field(min_length=1)


class SurfaceFallbacks(BaseModel):
    telegram: TelegramFallback | None = None
    slack: SlackFallback | None = None
    email: EmailFallback | None = None

    model_config = ConfigDict(extra="allow")  # forward-compatible for new channels


# --------------------------------------------------------------------------- #
# Top-level SurfaceSpec                                                       #
# --------------------------------------------------------------------------- #


class SurfaceSpecV1(BaseModel):
    schema_version: Literal["tilo.surface.v1"] = SURFACE_SCHEMA_VERSION
    surface_id: str = Field(min_length=1)
    turn_id: str = Field(min_length=1)
    run_id: str = Field(min_length=1)

    intent: SurfaceIntent
    budget_hint: BudgetHint
    block_compat: BlockCompat = BlockCompat.graceful

    blocks: Annotated[list[SurfaceBlock], Field(min_length=1)]
    fallback_text: str = Field(min_length=1)
    fallbacks: SurfaceFallbacks | None = None

    provenance: list[ProvenanceRef] = Field(default_factory=list)
    memory_refs: list[str] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)

    model_config = ConfigDict(use_enum_values=False)

    # ---- Cross-block validation (rules 7-8 of §7) ------------------------- #

    @model_validator(mode="after")
    def _validate_action_refs(self) -> "SurfaceSpecV1":
        seen_block_ids: set[str] = set()
        for block in self.blocks:
            if block.id in seen_block_ids:
                raise ValueError(f"Duplicate block id: {block.id!r}")
            seen_block_ids.add(block.id)

            available_action_ids = {a.id for a in block.actions}

            if block.type == SurfaceBlockType.decision:
                data = DecisionData.model_validate(block.data)
                self._check_action_refs(
                    block_id=block.id,
                    refs={opt.action_id for opt in data.options},
                    available=available_action_ids,
                    field="decision.option.action_id",
                )

            elif block.type == SurfaceBlockType.form:
                data = FormData.model_validate(block.data)
                self._check_action_refs(
                    block_id=block.id,
                    refs={data.submit_action_id},
                    available=available_action_ids,
                    field="form.submit_action_id",
                )

            elif block.type == SurfaceBlockType.editable:
                data = EditableData.model_validate(block.data)
                self._check_action_refs(
                    block_id=block.id,
                    refs={data.submit_action_id},
                    available=available_action_ids,
                    field="editable.submit_action_id",
                )

            elif block.type == SurfaceBlockType.artifact_link:
                data = ArtifactLinkData.model_validate(block.data)
                self._check_action_refs(
                    block_id=block.id,
                    refs={data.open_action_id},
                    available=available_action_ids,
                    field="artifact_link.open_action_id",
                )
        return self

    @staticmethod
    def _check_action_refs(
        *, block_id: str, refs: set[str], available: set[str], field: str
    ) -> None:
        for ref in refs:
            if ref == NOOP_ACTION_ID:
                continue
            if ref not in available:
                raise ValueError(
                    f"Block {block_id!r} {field} references missing action_id {ref!r}; "
                    f"available actions: {sorted(available) or '[]'}"
                )

    @field_validator("schema_version")
    @classmethod
    def _validate_schema_version(cls, value: str) -> str:
        if value != SURFACE_SCHEMA_VERSION:
            raise ValueError(f"Unsupported schema_version: {value!r}; expected {SURFACE_SCHEMA_VERSION!r}")
        return value


__all__ = [
    "SURFACE_SCHEMA_VERSION",
    "NOOP_ACTION_ID",
    "SurfaceIntent",
    "BudgetHint",
    "BlockCompat",
    "SurfaceBlockType",
    "Severity",
    "ProgressState",
    "FormFieldKind",
    "LinkTarget",
    "EditableKind",
    "DecisionMode",
    "ProgressShape",
    "HeadingData",
    "TextData",
    "EvidenceData",
    "ComparisonData",
    "ComparisonSide",
    "ComparisonRow",
    "DecisionData",
    "DecisionOption",
    "FormData",
    "FormField",
    "FormFieldOption",
    "ProgressData",
    "ProgressStep",
    "ListData",
    "ListItem",
    "LinkData",
    "EditableData",
    "ArtifactLinkData",
    "FallbackData",
    "SurfaceBlock",
    "SurfaceFallbacks",
    "TelegramFallback",
    "SlackFallback",
    "EmailFallback",
    "SurfaceSpecV1",
]
